persistence/internal/providertest/offsetstore/transaction.go

```diff
@@ -32,88 +32,68 @@ func DeclareTransactionTests(tc *common.TestContext) {
+								err := tx.SaveOffset(
+									tc.Context,
+									"<source-app-key>",
+									conflictingCurrentOffset,
+									nextOffset,
+								)
+								gomega.Expect(err).To(gomega.Equal(offsetstore.ErrConflict))
 
-					saveOffset(tc.Context, dataStore, "<source-app-key>", c, n)
+								return nil // let the transaction commit despite error
+							},
+						)
+						gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
 
-					actual, err := repository.LoadOffset(tc.Context, "<source-app-key>")
-					gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
-					gomega.Expect(actual).Should(gomega.BeNumerically("==", n))
-				})
+						actual, err := repository.LoadOffset(tc.Context, "<source-app-key>")
+						gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+						gomega.Expect(actual).To(gomega.BeEquivalentTo(0))
+					},
+					table.Entry("high pair low range", 1, 2),
+					table.Entry("high pair high range", 99, 100),
+				)
+			})
 
-				ginkgo.It("does not update the current offset with the equal next offset", func() {
+			ginkgo.When("application has previous offsets associated", func() {
+				ginkgo.It("updates the offset", func() {
 					c := offsetstore.Offset(0)
 					n := offsetstore.Offset(1)
 					saveOffset(tc.Context, dataStore, "<source-app-key>", c, n)
 
 					c = n
+					n += 123
 
 					saveOffset(tc.Context, dataStore, "<source-app-key>", c, n)
 
 					actual, err := repository.LoadOffset(tc.Context, "<source-app-key>")
 					gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
-					gomega.Expect(actual).Should(gomega.BeNumerically("==", n))
-				})
-			})
-
-			ginkgo.When("called multiple times within the same transaction", func() {
-				ginkgo.It("saves the last offset", func() {
-					c := offsetstore.Offset(0)
-					n := offsetstore.Offset(1)
-
-					err := persistence.WithTransaction(
-						tc.Context,
-						dataStore,
-						func(tx persistence.ManagedTransaction) error {
-							err := tx.SaveOffset(tc.Context, "<source-app-key>", c, n)
-							gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
-
-							c = n
-							n += n
-
-							err = tx.SaveOffset(tc.Context, "<source-app-key>", c, n)
-							gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
-
-							return nil
-						},
-					)
-
-					actual, err := repository.LoadOffset(tc.Context, "<source-app-key>")
-					gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
-					gomega.Expect(actual).Should(gomega.BeNumerically("==", n))
+					gomega.Expect(actual).To(gomega.BeEquivalentTo(n))
 				})
-			})
 
-			ginkgo.When("an OCC conflict occurs", func() {
 				table.DescribeTable(
-					"it does not update the offset",
+					"it does not update the offset when an OCC conflict occurs",
 					func(conflictingCurrentOffset, nextOffset offsetstore.Offset) {
-						// Update the current offset twice so that it's up to 2.
-						// Otherwise we can't test for 1 as a too-low value.
+						// Update the current offset twice so that it's up to 1.
```

Comment:
```suggestion
```

It's not being updated twice ;)

Reply:
Oh, I guess it is! sorry man, I think you can get away with a `BeforeEach()` here though that does the initial update from `0` for both this table test and the `updates the offset` test.

Reply:
So the both table tests (one in `application has no previous offsets associated` section, another in `application has previous offsets associated`) should not internal increment of the offset, right? But the one in `application has previous offsets associated` will rely on the incremented offset in `BeforeEach()`, correct? What about table entries should they remain as they are for both table tests?

Reply:
Regarding the offsets and `BeforeEach` yep that's what I meant, that should be neater.

I don't think the test in `application has no previous offsets associated` needs to be table-driven at all because there's no special case - any incorrect offset you give it is non-zero and too high. That's how the tests in #163 and the `queuestore` are structured.

For the `application has previous offsets associated` tests, you can bump "up past 1" in the single save done in `BeforeEach()`. Then I think you just need the `zero`, `too low` and `too high` test cases. There's probably no need to pass in the "new" offset in the table entries, it can just be hardcoded to some arbitrary high number (good ol' `123`) in the test function.